java/code/src/com/redhat/rhn/domain/user/UserFactory.java

```diff
@@ -703,30 +704,26 @@ public void setUserServerPreferenceValue(User user,
         return query.list();
     }
 
-    /**
-     * Return a list of all users.
-     *
-     * @return list of users.
-     */
-    public List<User> findAllUsers() {
-        CriteriaBuilder builder = getSession().getCriteriaBuilder();
-        CriteriaQuery<UserImpl> criteria = builder.createQuery(UserImpl.class);
-        criteria.from(UserImpl.class);
-        return getSession().createQuery(criteria).getResultList().stream()
-                .<User>map(Function.identity()).collect(Collectors.toList());
-    }
-
     /**
      * Return a list of all User's who are in the given org.
      *
-     * @param inOrg Org to find users for.
+     * @param inOrg Optional Org to find users for.
      * @return list of users.
      */
-    public List<User> findAllUsers(Org inOrg) {
-        Session session = HibernateFactory.getSession();
-        Map<String, Object> params = new HashMap<String, Object>();
-        params.put("org_id", inOrg.getId());
-        return listObjectsByNamedQuery("User.findAllUsersByOrg", params);
+    public List<User> findAllUsers(Optional<Org> inOrg) {
+        if (inOrg.isPresent()) {
+            Session session = HibernateFactory.getSession();
+            Map<String, Object> params = new HashMap<String, Object>();
+            params.put("org_id", inOrg.get().getId());
+            return listObjectsByNamedQuery("User.findAllUsersByOrg", params);
+        }
+        else {
+            CriteriaBuilder builder = getSession().getCriteriaBuilder();
+            CriteriaQuery<UserImpl> criteria = builder.createQuery(UserImpl.class);
+            criteria.from(UserImpl.class);
+            return getSession().createQuery(criteria).getResultList().stream()
```

Comment:
Nitpick - I wonder if we can just use criteria query only and add the where clause only if org is present.

Reply:
+1. On top of that, we are on a long-running effort to get rid of Criteria queries since they are deprecated.

We could keep the HQL query only. Something along the lines of:

```sql
WHERE (:org_id IS NULL) OR (:org_id = org_id)
```

Reply:
HQL and SQL will never really be obsolete! JPA queries are a different way, useful especially in very easy cases or when you need to compose them on the fly. Concatenating SQL strings is never fun.

Criteria queries occupied the same space as JPA queries and those have been deprecated.

Reply:
You mean something like
```java
    public List<User> findAllUsers(Optional<Org> inOrg) {
        String hql = "FROM com.redhat.rhn.domain.user.legacy.UserImpl as u " +
                "WHERE (:org_id IS NULL) OR (:org_id = org_id)";
        Query query = getSession().createQuery(hql);
        query.setParameter("org_id", Opt.fold(inOrg, () -> null, org -> org.getId()));
        return query.list();
    }
``` 
with inline HQL? 

I find it to be more readable, but it seems we are using named queries all over the place. Will this look good?

Reply:
That's is why I suggested to patch the existing query as follows:

```sql
WHERE (:org_id IS NULL) OR (:org_id = org_id)
```

 - if the parameter `:org_id` is `NULL` then return all users
 - otherwise only return users with matching `org_id`

Reply:
Ahhh, didn't know this trick . Yes, this works.

Reply:
Had some issues dealing with `(:org_id IS NULL)` (broken tests). Alternative can be:

Query
`from com.redhat.rhn.domain.user.legacy.UserImpl as u WHERE :org_id = -1 OR org_id = :org_id`

Code
```java
Map<String, Object> params = new HashMap<>();
int orgId = Opt.fold(inOrg, () -> -1, org -> org.getId().intValue());
params.put("org_id", orgId);
...
```

Reply:
I really wonder why it did not work. We can have a look together, the `-1` hack is not something I terribly like - if that's the only solution I would keep two separate queries for clarity.